[WebRTC] Make H265 packetizing more resilient to bad input

ddkilzer opened the pull request:
#### 7e7d41f82f4f250cf24d7f6b0e116ffefe42ef94
<pre>
[WebRTC] Make H265 packetizing more resilient to bad input
<a href="https://bugs.webkit.org/show_bug.cgi?id=265043">https://bugs.webkit.org/show_bug.cgi?id=265043</a>
&lt;<a href="https://rdar.apple.com/118460064">rdar://118460064</a>&gt;

Reviewed by Youenn Fablet.

Change a release assert to runtime failure to prevent easily hit
crashes.

Also improve performance of H265::FindNaluIndices() by making
H265::NaluIndex the same as H264::NaluIndex so there is no need to copy
identical data structures.

* Source/ThirdParty/libwebrtc/Source/webrtc/common_video/h265/h265_common.cc:
(webrtc::H265::FindNaluIndices): Remove.
- This code is no longer necessary because all it did was call
  H264::FindNaluIndices() and then translate H264::NaluIndex objects to
  H265::NaluIndex objects.  Changes in h265_common.h make this code
  obsolete.
* Source/ThirdParty/libwebrtc/Source/webrtc/common_video/h265/h265_common.h:
(struct H265::NaluIndex):
- Change to make identical to H264::NaluIndex with a using statement.
(webrtc::H265::FindNaluIndices):
- Change to an inline method that calls H264::FindNaluIndices().

* Source/ThirdParty/libwebrtc/Source/webrtc/modules/rtp_rtcp/source/rtp_format_h265.cc:
(webrtc::RtpPacketizerH265::RtpPacketizerH265):
- Fix typo that called H264::FindNaluIndices() instead of
  H265::FindNaluIndices().  Because of how H265::FindNaluIndices() was
  defined, this did not result in a bug, was likely a copy-paste
  mistake.
(webrtc::RtpPacketizerH265::GeneratePackets):
- Add missing checks for return values from PacketizeFu() and
  PacketizeSingleNalu().
(webrtc::RtpPacketizerH265::PacketizeSingleNalu):
- Change release assert into runtime check that returns early if
  fragment.size() is zero.

* Source/ThirdParty/libwebrtc/WebKit/0001-Make-H265-packetizing-more-resilient-to-bad-input.patch: Add.

Canonical link: <a href="https://commits.webkit.org/271214@main">https://commits.webkit.org/271214@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c036c8ab47501974e53d19b278bf396a60e88972

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/26768 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/5383 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/28009 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/28980 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/24473 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/7218 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/2778 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/24375 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/27030 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/4204 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/22980 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/3692 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/3734 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/23973 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/29464 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/24418 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/24373 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/29992 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/3770 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/1964 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/27897 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/5220 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/4239 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/3578 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/4123 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

xingri commented on Source/ThirdParty/libwebrtc/Source/webrtc/modules/rtp_rtcp/source/rtp_format_h265.cc:
```diff
@@ -201,7 +204,9 @@ bool RtpPacketizerH265::PacketizeSingleNalu(size_t fragment_index) {
+  if (fragment.size() == 0u) {
+    return false;
+  }
   packets_.push(PacketUnit(fragment, true /* first */, true /* last */,
```
@ddkilzer  Thanks for the catch. That rarely happens to my scenario but looks good to me.

ddkilzer replied:
Thanks @xingri.  Are you working on merging this code upstream?  If so, is there anything I can do to make that job easier?

Commit:
[WebRTC] Make H265 packetizing more resilient to bad input
https://bugs.webkit.org/show_bug.cgi?id=265043
<rdar://118460064>

Reviewed by Youenn Fablet.

Change a release assert to runtime failure to prevent easily hit
crashes.

Also improve performance of H265::FindNaluIndices() by making
H265::NaluIndex the same as H264::NaluIndex so there is no need to copy
identical data structures.

* Source/ThirdParty/libwebrtc/Source/webrtc/common_video/h265/h265_common.cc:
(webrtc::H265::FindNaluIndices): Remove.
- This code is no longer necessary because all it did was call
  H264::FindNaluIndices() and then translate H264::NaluIndex objects to
  H265::NaluIndex objects.  Changes in h265_common.h make this code
  obsolete.
* Source/ThirdParty/libwebrtc/Source/webrtc/common_video/h265/h265_common.h:
(struct H265::NaluIndex):
- Change to make identical to H264::NaluIndex with a using statement.
(webrtc::H265::FindNaluIndices):
- Change to an inline method that calls H264::FindNaluIndices().

* Source/ThirdParty/libwebrtc/Source/webrtc/modules/rtp_rtcp/source/rtp_format_h265.cc:
(webrtc::RtpPacketizerH265::RtpPacketizerH265):
- Fix typo that called H264::FindNaluIndices() instead of
  H265::FindNaluIndices().  Because of how H265::FindNaluIndices() was
  defined, this did not result in a bug, was likely a copy-paste
  mistake.
(webrtc::RtpPacketizerH265::GeneratePackets):
- Add missing checks for return values from PacketizeFu() and
  PacketizeSingleNalu().
(webrtc::RtpPacketizerH265::PacketizeSingleNalu):
- Change release assert into runtime check that returns early if
  fragment.size() is zero.

* Source/ThirdParty/libwebrtc/WebKit/0001-Make-H265-packetizing-more-resilient-to-bad-input.patch: Add.

Canonical link: https://commits.webkit.org/271214@main
